src/transformers/generation/flax_logits_process.py

```diff
@@ -455,3 +456,82 @@ def handle_cumulative_probs(logprobs_k, scores_k):
+
+    def __init__(self, ngram_size: int):
+        if not isinstance(ngram_size, int) or ngram_size <= 0:
+            raise ValueError(f"`ngram_size` has to be a strictly positive integer, but is {ngram_size}")
+        self.ngram_size = ngram_size
+
+    def get_previous_ngrams(self, input_ids: jnp.ndarray, vocab_size: int, cur_len: int):
+        """
+        get a matrix of size (batch_size,) + (vocab_size,)*n (for n-grams) that
+        represent the n-grams that occured previously.
+        The BCOO representation allow to store only the few non-zero entries, instead of the full (huge) matrix
+        """
+        batch_size, seq_len = input_ids.shape
+
+        all_update_indices = jnp.array(
+            [
+                [
+                    b,
+                ]
+                + [input_ids[b, i + j] for j in range(self.ngram_size)]
+                for i in range(seq_len - (self.ngram_size - 1))
+                for b in range(batch_size)
+            ]
+        )
+
+        data = jnp.ones((all_update_indices.shape[0],), dtype=jnp.uint16)
+        data = data * (jnp.arange(data.shape[0]) < batch_size * (cur_len - (self.ngram_size - 1)))  # ignore the n-grams not yet generated
+
+        return sparse.BCOO((data, all_update_indices), shape=(batch_size,) + (vocab_size,) * self.ngram_size)
+
+    def get_banned_tokens_mask(self, latest_tokens: jnp.ndarray, previous_ngrams) -> jnp.ndarray:
+        """
+        Determines which tokens must be banned given latest tokens and the previously seen
+        ngrams.
+        """
+
+        @sparse.sparsify
+        @jax.vmap
+        def inner_fn(latest_tokens, previous_ngrams):
+            vocab_size = previous_ngrams.shape[-1]
+            mask = jnp.ones((vocab_size,))
+            mask *= previous_ngrams[tuple(latest_tokens)]
+            return mask
+
+        return sparse.bcoo_todense(inner_fn(latest_tokens, previous_ngrams))
+
+    def __call__(self, input_ids: jnp.ndarray, scores: jnp.ndarray, cur_len: int) -> jnp.ndarray:
+        def true_fn():
+            _, vocab_size = scores.shape
+            # store the previously seen n-grams
+            previous_ngrams = self.get_previous_ngrams(input_ids, vocab_size, cur_len)
+
+            # get the n-1 last tokens that prefix the n-gram being generated
+            latest_tokens = jnp.zeros((input_ids.shape[0], self.ngram_size - 1), dtype=input_ids.dtype)
+            latest_tokens = jax.lax.dynamic_update_slice(
+                latest_tokens,
+                jax.lax.dynamic_slice(
+                    input_ids, (0, cur_len - (self.ngram_size - 1)), (input_ids.shape[0], (self.ngram_size - 1))
+                ),
+                (0, 0),
+            )
+
+            # compute the banned tokens, ie all the tokens that when added to the latest tokens lead to a n-gram that was previously generated
+            banned_tokens_indices_mask = jnp.invert(jnp.isclose(self.get_banned_tokens_mask(latest_tokens, previous_ngrams), 0))
```

Comment:
in order to not use a very large (and very sparse) matrix as `previous_ngrams`, i used the `sparse` module in jax. One issue I found with the `BCOO` data type, is that the `dtype` cannot be booleans (nor integers if i remember correctly).
You are right, these complicated casting operations might not be needed, i will check how to correctly cast to boolean from float, it will be more readable